android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt

```diff
@@ -176,3 +153,105 @@ sealed class RemoteSettingsResult {
      */
     data class NetworkFailure(val error: Exception) : RemoteSettingsResult()
 }
+
+private object JSONObjectSerializer : KSerializer<JSONObject> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("JSONObject", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: JSONObject) {
+        encoder.encodeString(value.toString())
+    }
+
+    override fun deserialize(decoder: Decoder): JSONObject {
+        val jsonString = decoder.decodeString()
+        return JSONObject(jsonString)
+    }
+}
+
+/**
+ * Data class representing serializable version of RemoteSettingsResponse.
+ */
+@Serializable
+data class SerializableRemoteSettingsResponse(
+    val records: List<SerializableRemoteSettingsRecord>,
+    val lastModified: ULong,
+)
+
+/**
+ * Data class representing serializable version of RemoteSettingsRecord.
+ */
+@Serializable
+data class SerializableRemoteSettingsRecord(
+    val id: String,
+    val lastModified: ULong,
+    val deleted: Boolean,
+    val attachment: SerializableAttachment?,
+    @Serializable(with = JSONObjectSerializer::class)
+    val fields: JSONObject,
+)
+
+/**
+ * Data class representing serializable version of (RemoteSettings) Attachment.
+ */
+@Serializable
+data class SerializableAttachment(
+    val filename: String,
+    val mimetype: String,
+    val location: String,
+    val hash: String,
+    val size: ULong,
+)
+
+private fun RemoteSettingsRecord.toSerializable(): SerializableRemoteSettingsRecord {
+    return SerializableRemoteSettingsRecord(
+        id = this.id,
+        lastModified = this.lastModified,
+        deleted = this.deleted,
+        attachment = this.attachment?.toSerializable(),
+        fields = this.fields,
+    )
+}
```

Comment:
After cycling on this several times, I'm sorry that we still ended up in a place with all this boilerplate. I'm personally the most happy with this result, but I think that this does indicate that there's a need for a better JSON bridge between us and A-S. 

Just to reframe our original issue, please correct me if my memory is wrong:

1. From A-S we got a fairly ergonomic type for use in Kotlin.
2. However, because we need to serialize that type so that we can save it to disk, we have to figure out a way to do that manually.
3. We tried using `JSONObject` to do this manually and found it to be a little messy, in that we had to extract every field using `getString`-type methods.
4. We tried using `kotlinx.serialization` to do it and ended up in a place where we have to write a lot of boilerplate to convert between `@Serializable` types.

Am I missing anything there?

Reply:
That's exactly it. Tbh, either way. `JSONObject` or Serialization looks like a lot of boilerplate code to me. 
My thinking? If we could have `JSONObject` being returned from the APIs instead of a new class i.e `RemoteSettingsResponse`?